[Relay] Merge analysis/context_analysis.cc and transforms/device_annotation.cc

mbs-octoml opened the pull request:
Currently LowerTEPass (backend/te_compiler.cc) is a 'special' pass because it
depends on a side-input DeviceMap. We'd like to remove that side-input, and
instead recover the Device (and, ultimately, Target) for each (fused) primitive
call from the AST alone.

By doing so we also avoid needing to perform device planning twice:
 - It needs to be done before lowering so we know which primitives need
   to be compiled for which devices.
 - It then needs to be re-done after lowering and optimization as a prelude
   to memory planning.
By baking the device plan into the AST we can simply do device planning before
lowering, and run memory planning later, both as ordinary passes.

While working on that issue we realized we currently have 3 'device planners':
 - transforms/device_annotation.cc, which supports only a small subset of Relay
   and uses a simple top-down algorithm to assign a device to every
   sub-expression.
 - analysis/context_analysis.cc, which makes a galant effort to support most of
   Relay, is based on unification rather than a top-down algorithm, but handles
   higher order functions by ad-hoc and fragile inlining.
 - transforms/annotate_target.cc, which works on Targets instead of Devices, but
   is otherwise like 'device planning'.
We'd like to bring these together.

In this PR we introduce a new transforms/device_planner.cc intended to replace
transforms/device_annotation.cc and analysis/context_analysis.cc. We don't
delete those two just yet since we need to switch all users off of them in the
next PR. We also leave transforms/annotate_target.cc alone pending a proper RFC
to bring devices and targets together sensibly, but have it firmly in our
sights.

transforms/device_planner.cc is based on analysis/context_analysis.cc, but
is heavily reworked to:
 1. Handle starting from existing "on_device" annotations as well as existing
    "device_copy" calls.
 2. Be idempotent, with the idea we'll probably need to re-run it to 'refine'
    device planning to account for storge scopes.
 3. Robustly handle all of Relay, particularly higher-order functions. For that
    we replace the inlining approach in analysis/context_analysis.cc with a
    higher-order unification domain.
 4. Be a little more systematic with defaulting.
 5. Capture the result of the analysis within the AST as new "device_copy" calls
    at device boundaries, and new/replaced "on_device" calls wherever the device
    for a sub-expression is not already 'obvious' from the sub-expression's
    lexical scope.
 6. Provide helper visitors for passes which need to ask for the device for
    any sub-expression they are processing and/or preserve device information
    on rewrites. Those passes include:
     - backend/aot_executor_codegen.cc (AOTOnDemandAllocator)
     - backend/graph_plan_memory.cc (StorageAllocaBaseVisitor etc)
     - backend/te_compiler.cc (LowerTensorExprMutator)
     - backend/vm/lambda_lift.cc (LambdaLifter)
     - transforms/memory_alloc.cc (DialectRewriter)
     - transforms/to_a_normal_form.cc (Fill)
     - backend/vm/compiler.cc (VMFunctionCompiler)
    However we won't change any of those in this PR.

See the draft #8788 for the end game, I'll be peeling PRs out of that.

mbs-octoml commented:
@electriclilies @mikepapadim @jroesch Here's part 1. Things to watch out for:
 - How to I may device_copy.cc branch from context_analysis.cc?
 - Did I go too far with device defaulting?
 - All the visitor helpers are needed in follow up but not used here, so sorry it's not obvious why I set things up like that. But suggestion for reducing code dup would be welcome.
 - I tried to make the intro comment comprehensive, let me know what needs more detail.
 - Anything non-obvious.
Thanks!

mbs-octoml commented:
Oh, and this is on top of #9012 which is still in-flight, sorry about that.

jroesch commented on tests/python/relay/test_pass_plan_devices.py:
```diff
@@ -0,0 +1,1405 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Now this is a test file!

jroesch replied:
Not to rain on your parade but it might be worth directly using text format here?

mbs-octoml replied:
Yeah it's pretty awful and I nearly switched a few times already. Which format should I use? Can it capture the attributes?

areusch replied:
+1 for relay text format, i think that would greatly help

mbs-octoml replied:
Done.

jroesch commented on src/relay/transforms/device_planner.cc:
```diff
@@ -0,0 +1,1986 @@
+ */
+
+/*!
+ * \file src/relay/analysis/device_planner.cc
```
I read everything except main file, will tackle this part tomorrow.

mbs-octoml replied:
Thanks. With apologies, but I ended up hoisting out everything except the main into https://github.com/apache/tvm/pull/9077, so I think you could directly LGTM that now?

I'm trying to model good PR hygiene :-)

areusch commented on src/relay/transforms/device_planner.cc:
```diff
@@ -0,0 +1,1986 @@
+ * Since higher-order aspects are later compiled away (by 'defunctionalization'
+ * aka 'firstification') we'd prefer not to have to answer any of those questions. In particular,
+ * we really don't want our domain \p D to allow for yet another device for the function closure.
+ * So we'll just force the 'device for a function' to be the same as the device for the function's
```
yeah this seems like a generally good design practice. people (or passes) can explicitly add copy nodes if that's what they want/need.

mbs-octoml replied:
Ack.

areusch commented on src/relay/transforms/device_planner.cc:
```diff
@@ -0,0 +1,1986 @@
+ * -------
+ *    ^
+ *    |
+ *    `-- Mark's stamp of completeness :-)
```
😂  will this become a new lint check?

areusch commented on src/relay/transforms/device_planner.cc:
```diff
@@ -0,0 +1,1986 @@
+ * We rewrite the programs to handle some special cases:
+ *  - "on_device" calls at the top-level of function or immediately let-bound are rewritten
+ *    to have \code is_fixed=true \endcode.
+ *  - We wish to treat \code on_device(expr, device_type=d).0 \endcode as if it were written
```
what's the difference here? are you saying we assume all results are tuples of length 1?

mbs-octoml replied:
No, it's just a clever peephole optimization hiding in the original device_annotation.cc that silently interprets a 'copy tuple then project' as if it were a 'project from tuple then copy'. It threw a bit of a spanner in the works and pushed me to require the 'phase 0' to spell it out explicitly since the trick in device_annotation.cc does not carry over into the unification-based approach in context_analysis.cc which this impl is derived from.

areusch commented on src/relay/transforms/device_planner.cc:
```diff
@@ -0,0 +1,1986 @@
+ * -------
+ * After flowing constraints we apply some defaulting heuristics (using a global default device)
+ * to fix the device for any as-yet unconstrained sub-expressions.
+ *  - Unconstrained function result devices default to the global default device.
```
should we instead default them to the most convenient device given the args?

areusch commented on src/relay/transforms/device_planner.cc:
```diff
@@ -0,0 +1,1986 @@
+ *  - Unconstrained function result devices default to the global default device.
+ *  - Unconstrained function parameters devices default to the device for the function result.
+ *  - Unconstrained let-bound expression devices default to the device for the overall let.
+ * TODO(mbs): I may have over-innovated here and we simply want to bind all free domaints to
```
actually my feeling is that there may (later) be some desire for user control over this. like a fuzzy device constraint or optimization target. e.g. "Optimize for fewer copies at the expense of performance" because we might not quite get the memory bus modelling right for quite some time yet (or ever, depending on which SoC is being worked with).

areusch commented on src/relay/transforms/device_planner.cc:
```diff
@@ -0,0 +1,1986 @@
+      if (lhs->device_type_ != rhs->device_type_) {
+        // TODO(mbs): Proper diagnostics.
+        std::ostringstream os;
+        os << "Inconsistent device types " << lhs->device_type_ << " and " << rhs->device_type_;
```
shall we print more details somehow, like the sub-expression, or use Jared's expression annnotating infra? or maybe this does and i'm just dense :)

jroesch replied:
Yeah I think we can do this, just need to do some plumbing, Mark and I can add tracking for this as part of turning this thing online.

mbs-octoml replied:
Filed CORE-75. Yeah, I'm up for getting the diagnostics to a usable state -- when things went wrong on my unit tests it was tough going figuring out the root cause.

areusch commented on src/relay/transforms/device_planner.cc:
```diff
@@ -0,0 +1,1986 @@
+   * callee is a primitive or special operation we handle it specially. Otherwise defers to \p
+   * DomainFor(call->op).
+   *
+   * This special handling is needed:
```
this smells a bit like an overridable Pass...wdyt?

jroesch replied:
We could add a hook here, but I would like to first make sure everyone is using the same algorithm before we provide anymore customization.

mbs-octoml replied:
Agree - this at least gets things into a state where we could do that, or allow a table to be passed in, or something. Most of this is a large refactor from the VisitExpr_(CallNode*) in context_analyis.cc, but without needing to repeat all the args handling. I was actually happy with that cleanup and want to lock that in :-)

areusch commented on src/relay/transforms/device_planner.cc:
```diff
@@ -0,0 +1,1986 @@
+/*!
+ * \brief Rewrites "on_device" calls to handle some special cases.
+ */
+class RewriteOnDevices : public ExprMutator {
```
dang this file is getting a bit long..

jroesch replied:
Yeah might be worth splitting out into like device_domain.{h, cc} and then putting the core pass parts in their own file.

areusch commented on src/relay/transforms/device_planner.cc:
```diff
@@ -0,0 +1,1986 @@
+      domains_->Unify(func_domain, implied_domain);  // higher-order
+    } catch (const Error& e) {
+      // TODO(mbs): Proper diagnostics.
+      LOG(FATAL) << "Function parameters and result devices do not match those of call. Call:"
```
!! fatal! probably we need to address the TODO before merging...i think?

mbs-octoml replied:
Yeah this is a big improvement on the context_analysis.cc (which will CHECK fail with no context). Let's do this as CORE-75.

areusch commented on src/relay/transforms/device_planner.cc:
i roughly understand what's happening, but i think example Relay snippets would help to better understand/maintain each Pass here.

mbs-octoml replied:
I added some simple examples to each of the internal pass class comments.

areusch commented on tests/python/relay/test_pass_plan_devices.py:
```diff
@@ -0,0 +1,1405 @@
+
+
+if __name__ == "__main__":
+    test_plain()
```
`sys.exit(pytest.main([__file__] + sys.argv[1:]))`

one day we will write a helper :)

jroesch replied:
ping pong on this one

mbs-octoml replied:
done

areusch reviewed (commented):
@mbs-octoml did an initial pass, don't think i understand everything yet, but i can see the pieces here. thanks for doing this!!

mbs-octoml commented:
While converting the unit tests to script form I ended up needing #9130.

I'll work through the remaining comments now.

jroesch commented on src/relay/transforms/device_planner.cc:
```diff
@@ -0,0 +1,1994 @@
+ *    incorporate targets and memory scopes into the domain. For example it's ok for the function
+ *    body to be executed on different device ids provided they have the same target and memory
+ *    scope.
+ *  * Might be simpler to just let every type have a device annotation rather than work in
```
cc @areusch

jroesch reviewed (approved):
I think everything makes sense, big asks would to be splitting up the passes into smaller files with a better separation of the pieces and adding diagnostics and other features. I am totally OK landing this version and building action item list to do before we turn this on as default pass.

mbs-octoml commented on python/tvm/relay/op/annotation/annotation.py:
```diff
@@ -33,21 +43,26 @@ def on_device(data, device):
+
+
+# for testing only
+def function_on_device(function, param_devices, result_device):
```
I ended up fully commenting this one in a prior PR.

mbs-octoml commented on src/relay/op/annotation/annotation.cc:
```diff
@@ -36,15 +38,40 @@
+  return Call(OnDeviceOp(), {std::move(expr)}, Attrs(std::move(attrs)), /*type_args=*/{}, span);
+}
+
+Expr OptOnDevice(Expr expr, DLDeviceType device_type, bool is_fixed) {
```
Ah, old personal naming convention but it clashes with Optimize. Switching to Maybe (in the supporting PR).

mbs-octoml commented on src/parser/parser.cc:
```diff
@@ -1088,7 +1088,7 @@ class Parser {
     return WithSpan<Function>([&]() {
       PushScope();
-      PushTypeScope();
+      PushTypeScope();  // TODO(mbs): BUG?
```
I think the second PushTypeScope() is in error and I'll delete it. You unconditionally pop one scope further down.

mbs-octoml commented on src/relay/op/annotation/annotation.cc:
```diff
@@ -128,24 +128,31 @@ OnDeviceProps GetOnDeviceProps(const Expr& expr) {
-  attrs->result_device_type = result_device_type;
-  return WithAttr(std::move(function), attr::kFunctionAttrsKey, Attrs(std::move(attrs)));
+                          Integer result_device_type) {
+  return WithAttr(WithAttr(std::move(function), tvm::attr::kParamDeviceTypes, param_device_types),
```
Done.

mbs-octoml commented on src/relay/op/annotation/annotation.cc:
```diff
@@ -56,12 +83,98 @@ RELAY_REGISTER_OP("on_device")
 
+OnDeviceProps GetOnDeviceProps(const CallNode* call_node) {
+  if (call_node->op == OnDeviceOp()) {
+    ICHECK_EQ(call_node->args.size(), 1) << "on_device expects one argument";
```
I've gotten all the way through to unit testing the final PR without needing it, so how about I leave it. I tend to add them as I need them.

mbs-octoml commented on src/relay/op/annotation/annotation.h:
```diff
@@ -0,0 +1,116 @@
+ */
+OnDeviceProps GetOnDeviceProps(const Expr& expr);
+
+/*! \brief Returns true if \p expr is an on_device CallNode. */
```
Turns out this is dead code now so nuked.

mbs-octoml commented on src/relay/op/memory/device_copy.cc:
```diff
@@ -0,0 +1,117 @@
+
+RELAY_REGISTER_OP("device_copy")
+    .describe(R"code(
+Copy data from one tensor to another. The source and destination might be
```
This should show up as a branch from memory.cc, sorry. The operator is not new, and can appear is user programs.

mbs-octoml commented:
@areusch, @jroesch thank you and PTAL:
 - Rebased on main
 - All unit tests (plus a new one for match :-) in script form
 - Split out devices domains and device-aware visitors into separate files
 - Cpp unit test for device domains to prompt us to better support them
 - Addressed above comments.

jroesch commented:
I am going to land this, if people have follow-ups we can do them in the final PR which actually turns this on.
